Add support for replacing dependencies

mprahl opened the pull request:
To accomplish this, fetch_app_source now extracts the application
source code archive to the `<bundle_dir>/temp/<request_id>/app`
directory. This can then be used by the package manager tasks to
modify the source for dependency overrides. For Golang, this ends
up being the go.mod and go.sum files.

The create_bundle_archive task then creates an archive from the
contents from `<bundle_dir>/temp/<request_id>/app` and
`<bundle_dir>/temp/<request_id>/deps`.

The dependencies in a request's JSON now has a new key called
"replaces". If this is set, a JSON object representing the
dependency that was replaced is shown.

yashvardhannanavati commented on cachito/web/models.py:
```diff
@@ -162,6 +175,18 @@ def number_of_deps(self):
             req_dep_cols.request_id == self.id).scalar()
 
+    @property
+    def number_of_dep_overrides(self):
```
How about renaming this to dependency_overrides_count

mprahl replied:
@yashvardhannanavati I'm just using that name to be consistent with `number_of_deps`. What's funny is you named it that way in #52.  :laughing:

If I rename this one, I'll also rename the other. Which do you prefer?

yashvardhannanavati replied:
@mprahl lol, Sorry I didn't realize it while writing #52. Can I please request you to rename both if possible?

yashvardhannanavati commented on cachito/web/models.py:
```diff
@@ -242,6 +271,22 @@ def from_json(cls, kwargs):
+            raise ValidationError('"dependency_overrides" must be an array')
+
+        dependency_override_objects = []
+        for dep in dependency_overrides:
```
I prefer having the entire name `dependency` instead of `dep`.

I blame luiz's book for that. :p

yashvardhannanavati replied:
But I'll leave it up to you :)

mprahl replied:
Sure, I'll use `dependency_override` to be more descriptive.

yashvardhannanavati commented on cachito/workers/pkg_manager.py:
```diff
@@ -70,6 +79,11 @@ def resolve_gomod_deps(archive_path, request_id):
             if len(parts) == 1:
                 # This is the application itself, not a dependency
                 continue
+            elif len(parts) == 3:
```
What are your thoughts on converting this piece into a switch statement instead? I think it would look cleaner.

mprahl replied:
@yashvardhannanavati Python doesn't support switch statements.

yashvardhannanavati replied:
Ah, that's right. My bad. Java is playing with my mind again.

yashvardhannanavati commented on cachito/workers/pkg_manager.py:
```diff
@@ -148,31 +162,17 @@ def add_deps_to_bundle(src_deps_path, dest_cache_path, request_id):
     """
-    config = get_worker_config()
-    deps_path = os.path.join(config.cachito_bundles_dir, 'temp', str(request_id), 'deps')
+    bundle_dir = get_request_bundle_dir(request_id)
```
Optional: You can skip creating this variable here and directly call the method in the next statement.

mprahl replied:
I did it the other way for readability, but I'll change it.

mprahl commented:
@yashvardhannanavati I addressed your comments. Could you take another look?

lcarva commented on cachito/workers/pkg_manager.py:
```diff
@@ -61,19 +61,27 @@ def resolve_gomod_deps(archive_path, request_id=None):
-            if len(parts) > 2:
-                log.warning('Unexpected go module output: %s', line)
-                continue
+            elif len(parts) == 4:
```
Consider using `if` instead of `elif` here. This seems to create some sort of relationship with the previous `if` statement that I don't think is accurate.

lcarva commented on tests/test_workers/test_pkg_manager.py:
```diff
@@ -22,7 +22,7 @@
     github.com/kr/text v0.1.0
     github.com/op/go-logging v0.0.0-20160315200505-970db520ece7
-    github.com/pkg/errors v0.8.1
+    github.com/pkg/errors v0.8.0 github.com/pkg/errors v0.8.1
```
It would be nice to have a test case that verified v0.8.1 is used. Also, to ensure that the second module name is used, maybe change the line to something like:
```
github.com/pkg/ignored v0.8.0 github.com/pkg/errors v0.8.1
```

I don't think we need to do it for every dep here, but a few keys one would be useful.

mprahl replied:
This actually does test that `v0.8.1` is used. I can add a test to for the different name though.

lcarva commented on tests/test_workers/test_tasks.py:
```diff
@@ -34,34 +34,29 @@ def test_fetch_app_source_request_timed_out(mock_git, mock_set_request_state):
-    (None, True, False),
+@pytest.mark.parametrize('auto_detect, contains_go_mod', (
+    (False, False),
+    (False, False),
```
Duplicated test case.

lcarva commented on tests/test_workers/test_tasks.py:
```diff
@@ -34,34 +34,29 @@ def test_fetch_app_source_request_timed_out(mock_git, mock_set_request_state):
-    )
-    if request_id_to_update:
+    tasks.fetch_gomod_source(app_archive_path, 1, auto_detect=auto_detect)
+    if not auto_detect or contains_go_mod:
```
Consider adding a new test param `expect_state_update` instead.

lcarva commented on cachito/web/api_v1.py:
```diff
@@ -125,12 +125,21 @@ def create_request():
+        ]
         chain_tasks.append(
-            tasks.fetch_gomod_source.s(request.id, auto_detect=auto_detect).on_error(error_callback)
+            tasks.fetch_gomod_source.si(
```
+1 to using immutable signatures. I think this makes the code easier to understand.

lcarva commented on cachito/web/models.py:
```diff
@@ -162,6 +175,18 @@ def number_of_deps(self):
+        """
+        req_dep_cols = request_dependency_override_table.c
+        return db.session.query(sqlalchemy.func.count(req_dep_cols.dependency_id)).filter(
+            req_dep_cols.request_id == self.id).scalar()
```
Didn't know about [scalar](https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.scalar). We should use it more often.

lcarva commented on cachito/web/models.py:
```diff
@@ -197,15 +222,19 @@ def to_json(self, verbose=True):
             rv.update({'state_history': states})
             rv.update({'dependencies': [dep.to_json() for dep in self.dependencies]})
+            rv.update(
+                {'dependency_overrides': [dep.to_json() for dep in self.dependency_overrides]}
```
Hmm... does this mean the `dependencies` attribute is not an accurate representation of the dependencies being used?

lcarva replied:
It would be more useful if the existing items in the `dependencies` list had additional information to determine whether or not they are overrides. That way clients only have to consume a single list, but still have the overrides information if needed.

mprahl replied:
The `dependencies` attribute is accurate. It's the dependencies that golang will use during a build, which includes the replace directives in the `go.mod` file.

I'll reach out on IRC on how we want to approach the comment on the JSON representation.

lcarva commented on cachito/workers/pkg_manager.py:
```diff
@@ -34,20 +34,23 @@ def __exit__(self, exc, value, tb):
-    :param str archive_path: the full path to the application source code
+    :param str app_source_path: the full path to the application source code
     :param int request_id: the request ID of the bundle to add the gomod deps to
+    :param list dep_overrides: dependency overrides with the keys "name" and "version; this
```
Missing double quote after `version`.

lcarva commented on cachito/workers/pkg_manager.py:
```diff
@@ -56,7 +59,13 @@ def resolve_gomod_deps(archive_path, request_id):
+            name = dep_override['name']
+            version = dep_override['version']
+            log.info('Applying the gomod override %s@%s', name, version)
+            _run_cmd(('go', 'mod', 'edit', '-replace', f'{name}={name}@{version}'), run_params)
```
I think we want to have the ability to override across different names. This would be needed for handling replacing a dependency with a fork done in another host, for example. `github.com/foo/bar` gets replaced with `git.internal.com/foo/bar`.

mprahl commented:
@lcarva and @yashvardhannanavati this is ready for review again.

mprahl commented on cachito/web/models.py:
```diff
@@ -143,6 +236,36 @@ class Request(db.Model):
+            # Send the changes queued up in SQLAlchemy to the database's transaction buffer
+            db.session.flush()
+
+        mapping = RequestDependency.query.filter_by(
```
An exception is raised if this occurs now. I also added a test for this.

mprahl commented:
@lcarva I addressed your comments. Here is the diff for your convenience:
https://github.com/release-engineering/cachito/compare/2afeb0f4829b179129337ac21dab41af90b07d09..a8822bcd6512f194112fc800ed4aaa668d8a39a0
